In [31]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [32]:
df_train = pd.read_csv("train_KNN.csv")
df_test = pd.read_csv("test_KNN.csv")

In [33]:
df_train = df_train.dropna(subset=["Embarked"])
df_train["Age"] = df_train["Age"].fillna(df_train["Age"].median())
df_train = df_train.drop(["Cabin"], axis=1)

df_test = df_test.dropna(subset=["Fare"])
df_test["Age"] = df_test["Age"].fillna(df_test["Age"].median())
df_test = df_test.drop(["Cabin"], axis=1)

In [34]:
df_train.drop(["PassengerId","Name","Ticket"], axis=1, inplace=True)
df_test.drop(["PassengerId","Name","Ticket"], axis=1, inplace=True)

In [35]:
df_train["Survived"] = df_train["Survived"].astype(str)
df_train["Pclass"] = df_train["Pclass"].astype(str)

df_test["Survived"] = df_test["Survived"].astype(str)
df_test["Pclass"] = df_test["Pclass"].astype(str)

In [36]:
df_train_num = df_train.select_dtypes(exclude="object")
df_train_obj = df_train.select_dtypes(include="object")

df_test_num = df_test.select_dtypes(exclude="object")
df_test_obj = df_test.select_dtypes(include="object")

df_train_obj = pd.get_dummies(df_train_obj, drop_first=True)
df_test_obj = pd.get_dummies(df_test_obj, drop_first=True)

In [37]:
Final_train_df = pd.concat([df_train_num, df_train_obj, df_train["Survived"]], axis=1)
Final_test_df = pd.concat([df_test_num, df_test_obj, df_test["Survived"]], axis=1)

In [38]:
X_train = Final_train_df.drop("Survived", axis=1)
y_train = Final_train_df["Survived"]

X_test = Final_test_df.drop("Survived", axis=1)
y_tes = Final_test_df["Survived"]

In [39]:
scaler = StandardScaler()
scaled_X_train = scaler.fit_transform(X_train)
scaled_X_test = scaler.transform(X_test)

In [40]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(scaled_X_train, y_train)

y_pred = knn.predict(scaled_X_test)

print("KNN Accuracy:", accuracy_score(y_tes, y_pred))

KNN Accuracy: 0.9928057553956835


In [41]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(max_iter=1000)
lr.fit(scaled_X_train, y_train)

y_pred_lr = lr.predict(scaled_X_test)

print("Logistic Regression Accuracy:", accuracy_score(y_tes, y_pred_lr))

Logistic Regression Accuracy: 1.0


In [42]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier()
dt.fit(scaled_X_train, y_train)

y_pred_dt = dt.predict(scaled_X_test)

print("Decision Tree Accuracy:", accuracy_score(y_tes, y_pred_dt))

Decision Tree Accuracy: 1.0


In [43]:
print("\nModel Comparison:")
print("KNN:", accuracy_score(y_tes, y_pred))
print("Logistic Regression:", accuracy_score(y_tes, y_pred_lr))
print("Decision Tree:", accuracy_score(y_tes, y_pred_dt))


Model Comparison:
KNN: 0.9928057553956835
Logistic Regression: 1.0
Decision Tree: 1.0


Conclusion:

The KNN model achieved a high accuracy of approximately 99.28%.

Both Logistic Regression and Decision Tree achieved perfect accuracy of 100%, outperforming KNN.

This indicates that these models were able to capture the patterns in the dataset more effectively.

However, such perfect accuracy may also indicate possible overfitting.

Overall, Logistic Regression and Decision Tree performed the best for this dataset.